### PR TITLE
fix repository update with the same settings but different type

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: Maybe we can just use RepositoryMetaData.equals() here?

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java]: You can use MockRepository.Plugin instead of adding a new type of repository

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java]: Why not using randomPath() directly?

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java]: I'm wondering if we should also check the Repository instance itself using internalCluster().getInstance(RepositoriesService.class).repository(name) and check if the instance is the same or has changed.

**Review by tlrx — COMMENTED** [review]
I left some comments

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java
Comment (vladimirdolzhenko) [reply server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java]: agree - make sense to add that kind of check

**Comment by vladimirdolzhenko** [comment]
@tlrx thanks for your comments, could you pls have another look ?

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: You can use assertAcked(client.admin().cluster().preparePutRepository(repositoryName)
                .setType(FsRepository.TYPE)
                .setSettings(repoSettings))

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: Maybe we could make this more readable with:
`final String updatedRepositoryType = randomBoolean() ? "mock" : FsRepository.TYPE`

?

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: You can use `assertAcked()`

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: I don't think we need to check the implementation class, instance should be enough.

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: Can you use more explicit name? (Also for getRepositoriesResponse1/2).

**Review by tlrx — COMMENTED** [review]
I left more minor comments, thanks @vladimirdolzhenko for fixing this!

**Comment by vladimirdolzhenko** [comment]
test this please

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: Can you make this final? Also, I think you should use `getDataOrMasterNodeInstances` as the repository is only registered on master eligible and data nodes. The method `getInstance()` retrieves the instance from a random node and if it's not a data or master one it will not find the repository.

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: If it's not used afterwards, maybe we don't need a ref to the location. We could simply use `.put("location", randomRepoPath())`

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: Nit: typo in originalRetRepositoriesResponse

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: Can you make it final?

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: `updated` and `updatedRepositoryType` can be final too

**Review by tlrx — COMMENTED** [review]
Another bunch of minor comments, sorry I didn't catch everything the first times.

**Comment by vladimirdolzhenko** [comment]
test this please

**Comment by vladimirdolzhenko** [comment]
test this please

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: You don't need to pass the default cluster `settings` here but `location` is enough for a FS repository

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: I think you can remove this comment

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java]: Thinking about it twice, I don't think that this is necessary. I think you can remove it, run the test multiple times andsee if it fails or not.

**Review by tlrx — APPROVED** [review]
LGTM - I left a more comments that I'd like to be addressed before merging but I don't think that it will require another review

**Pull request merged**
